### DISCUSSION TITLE
168

### DISCUSSION THREAD

**Discussion by jojoldu** [jojoldu/blog-comments#724] [Comments]
http://jojoldu.tistory.com/168

**Comment by jojoldu** [comment]
작성자: 박수영

동욱님 안녕하세요! 글 잘 읽었습니다 :) 
읽다 보니 저도 최근에 비슷한 업무를 하고 있어서요!! 포스트를 신나게 읽었습니다 ㅋㅋ 
다름이 아니라 질문 하나 하고 싶은데요! ㅎㅎ
관리자 사이트가 여러개 있고, 인증/권한을 한 시스템에서 관리하는것을 스프링 시큐리티로 해결할수가 있나요? 음.. 다른 도메인을 가진 시스템에서 인증되었다고 알수 있는 방법이 있나요? 

(이글에 답글이 달리면 알림은 안주겠죠..?ㅋㅋ 잘 확인하겠습니다!)

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

안녕하세요 수영님! ㅋㅋ 오랫만이에요 ㅎㅎ
말씀하신 방법은 해당 백오피스(관리자 사이트)가 인증권한 시스템의 DB에 접근할수있다면 가능할것 같습니다.
인증권한 사이트에서 회원의 인증, 권한, 그룹관리 등을 지정하고
다른 백오피스에서 그 DB를 시큐리티의 인증권한 DB로 사용하면 해결될것 같습니다.

**Comment by jojoldu** [comment]
작성자: Janghee Lee

창천향로님 좋은 글 감사합니다!! 지금 진행하고 있는 팀프로젝트가 Spring-boot에서 Security를 활용한 소셜로그인 기능이 필요했는데 spring.io의 예제보다 훨씬 좋은 것같습니다.(제가 영어번역기로 예제를 따라하다 오류를 잡는 속도가 발생하는 속도를 못따라가서..)

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

앗 그렇게나 좋게 봐주시다니 ㅠ
제가더 감사드립니다.
좀 더 상세하고 친절하게 하나씩 올리겠습니다.
감사합니다!

**Comment by jojoldu** [comment]
작성자: 행인

혹시 csrf 를 활성화 시키지 않은 이유가 있는지 궁금합니다..

**Comment by jojoldu** [comment]
작성자: 창천향로 창천향로

안녕하세요!
csrf를 끈 이유는 2가지인데요
1) 로컬환경에서 H2 Console에 접근하기 위해
2) 불필요한 테스트 코드 작성을 줄이기 위해
입니다.

실제로 프로젝트 진행시에도 profile이 local일 경우에는 csrf를 off하고, dev, production 일때는 on 하는 방식으로 구현합니다.

코드 참고:
https://www.slipp.net/questions/546#answer-1666

**Comment by brewagebear** [comment]
궁금한 점이 있어서 남겨봅니다.

최근에 집필하신 스프링부트와 AWS로 혼자 구현하는 웹 서비스 책을 잘 읽었습니다.

해당 OAuth2 구현 부분에서 문제에 부딪혔는데 기존 예제 소스는 OAuth2UserService를 구현체로 하여 OAuth2UserRequest를 활용하여 해당 OAuth로 로그인 처리 된 후 유저 값을 가져와 DB에 넣는 예제로 알고있습니다.

여기까지는 금방 따라했으나, Access Token이 아닌 Refresh Token을 DB에 넣으려고하다보니 문제가 발생했습니다. 

제가 디버깅을 해보면서 추적한 결과 `OAuth2LoginAuthenticationFilter`로 저희가 만든 Provider의 정보로 들어온 후에 `attemptAuthentication()` 함수를 통하여 `this.authorizedClientRepository.saveAuthorizedClient(authorizedClient, oauth2Authentication, request, response);` 처리를 통하여 인증정보를 저장하는 것으로 보입니다.

하지만, `OAuth2UserRequest`에 refresh token 값이 없었습니다. 그러나, 위에 라인을 디버깅을 했을 경우 refresh token이 제대로 나왔습니다. 제 생각에는 애초에 OAuth2UserRepuest의 생성자 자체가 Access Token만 가지고 있어서 그런걸로 보입니다. 

이때 가진 의문은 아래와 같습니다.
1. Refresh Token을 DB에 저장하기 위해서 OAuth2UserService가 아닌 OAuth2AuthorizedClientService를 구현하는게 맞는건지? 

`OAuth2AuthorizedClientService`는 `OAuth2AuthorizedClient`를 갖기때문에 해당 인자에서 바로 리프레시 토큰을 뽑을 수 있는 것으로 압니다. 하지만, 이 구현체는 UserEndpoint에 사용하기에는 적합하지 않다라고 알고 있습니다. 애초에 구현체가 `OAuth2UserService`인 놈만 인자로 받으니까요.

**그렇다면, 스프링 프레임워크는 User 엔티티에 리프레시 토큰을 저장하는 걸 지양하는 건가요?**

그래서 아래의 의문이 생겼습니다.

2. 사용자 테이블에 리프레시 토큰 값을 넣는게 맞는건지? (REST API입니다.) 

제가 사용하는 로그인 서비스는 카카오 로그인입니다.
액세스 토큰의 만료일이 다가오거나 할 경우 다시 토큰을 재발급 하기위해서는 리프레시 토큰이 필요하게끔 설계가 되어있는 것으로 압니다. 

따라서 토큰 재갱신이 필요할 때는 어떻게 처리를 해야되는지 궁금합니다.
아예 expired_at()이 다가오면 새로운 토큰을 발급하는게 맞는걸까요?


블로그나 책이나 항상 양질의 정보를 주셔서 감사합니다.

**Comment by her0807** [comment]
안녕하세요 동욱님!

동욱님께서 내신 책을 보고 코린이로 성장중인 백엔드 개발자를 희망하는 학생입니다! ㅎㅎ

다름이 아니라, 예제에 사용하신 OAuth2ClientContext 이 현재는 사용 불가하여 

대체할만한 예제를 만들어주시면 좋겠습니다. 

제가 현재 개발하고 있는 서비스에 소셜 로그인을 적용했는데, 

참고할만한 테스트 코드가 없어서 애를 먹고 있습니다 .. ㅠㅠㅠㅠ

**Comment by jyeonjyan** [comment]
**her0807님 대신 도움을 드리겠습니다!**
> 다름이 아니라, 예제에 사용하신 OAuth2ClientContext 이 현재는 사용 불가하여

네 동욱님이 이 글을 작성하신 날짜는 `2017. 8. 5.` 이고,
`2017.10`에 springboot 2.x 버전대에 변화가 생겨서 그런겁니다!

결론적으로 spring security oauth 프로젝트의 일부 기능이 spring security 로 마이그레이션 되는 중이라 기존의 oauth 프로젝트로는 지원이 안되는 기능이 생긴건데요.

아래 속성을 build.gradle 을 추가해주시면 `v1.x`에 만들어 둔 것들에 호환이 가능할거예요!
```gradle
repositories {
	mavenCentral()
	/** spring 2.x 버전과 1.x 버전의 호완성을 위해 **/
	maven { url 'https://repo.spring.io/snapshot' }
}
dependencies {
         /** @EnableOAuth2Client 등을 사용하기 위함 **/
	implementation 'org.springframework.security.oauth.boot:spring-security-oauth2-autoconfigure:2.0.1.BUILD-SNAPSHOT'
}
```

도움이 되었으면 좋겠습니다. 감사합니다.
